test/pylib/resource_gather.py

```diff
@@ -0,0 +1,195 @@
+        os.setsid()
+
+    def get_test_metrics(self) -> Metric:
+        pass
+
+    def write_metrics_to_db(self, metrics: Metric, success: bool = False) -> None:
+        pass
+
+    def cgroup_monitor(self, test_event: Event):
+        pass
+
+
+class ResourceGatherOff(ResourceGather):
+    def cgroup_monitor(self, test_event) -> Task:
+        return asyncio.create_task(no_monitor())
+
+
+class ResourceGatherOn(ResourceGather):
+    def __init__(self, test, tmp_dir: str):
+        super().__init__(test, tmp_dir)
+        self.sqlite_writer = SQLiteWriter(self.db_path)
+        self.test_id: int = self.sqlite_writer.write_row_if_not_exist(
+            Test(
+                architecture=platform.machine(),
+                directory=test.suite.name,
+                mode=test.mode,
+                run_id=test.id,
+                test_name=test.shortname
+            ),
+            TESTS_TABLE)
+
+    def make_cgroup(self):
+        os.makedirs(self.cgroup_path, exist_ok=True)
+
+    def get_test_metrics(self) -> Metric:
+        test_metrics: Metric = Metric(test_id=self.test_id)
+        test_metrics.time_taken = round(int(self.test.time_end - self.test.time_start))
+        test_metrics.time_start = datetime.fromtimestamp(self.test.time_start).strftime(DATE_TIME_TEMPLATE)
+        test_metrics.time_end = datetime.fromtimestamp(self.test.time_end).strftime(DATE_TIME_TEMPLATE)
+        test_metrics.success = self.test.success
+        with open(self.cgroup_path / 'memory.peak', 'r') as file:
+            test_metrics.memory_peak = file.read()
+
+        if (self.cgroup_path / 'cpu.stat').exists():
+            with open(self.cgroup_path / 'cpu.stat', 'r', ) as file:
+                self._parse_cpu_stat(file, test_metrics)
+        return test_metrics
+
+    def write_metrics_to_db(self, metrics: Metric, success: bool = False) -> None:
+        metrics.success = success
+        self.sqlite_writer.write_row(metrics, METRICS_TABLE)
+
+
+    def put_process_to_cgroup(self):
+        super().put_process_to_cgroup()
+        pid = os.getpid()
+        with open(self.cgroup_path / 'cgroup.procs', "a") as cgroup:
+            cgroup.write(str(pid))
+
+    def cgroup_monitor(self, test_event: Event) -> Task:
+        return asyncio.create_task(self._monitor_cgroup(test_event))
+
+    async def _monitor_cgroup(self, test_event: Event) -> None:
+        """Continuously monitors CPU and memory utilization."""
+        while not test_event.is_set():
+            with  open(self.cgroup_path / 'memory.current', 'r') as memory_current:
+                timeline_record = CgroupMetric(
+                    test_id=self.test_id,
+                    memory=int(memory_current.read()),
+                    timestamp=datetime.now()
+                )
+
+                self.sqlite_writer.write_row(timeline_record, CGROUP_MEMORY_METRICS_TABLE)
+
+            # Control the frequency of updates, for example, every 2 seconds
+            await asyncio.sleep(1)
```

Comment:
Since you have the memory.peak file, any reason to have this coroutine?

Reply:
For better understanding of the memory use. For example, we have one test in a file that consumes a lot of memory, but the rest of the test do not need so much memory. With only peak memory, we will lose this information, and we will not be able to schedule the test correctly.

Reply:
I don't think we'll be able to utilize this information. The execution history depends on what other tests ran in parallel. The next execution will have different timing.

Reply:
Our tests executed in alphabetical order. So every run has the same order of execution. Until we start shuffling them, we can compare this data between runs.